### ISSUE TITLE
A Proposal for Handling Generic Events

### ISSUE THREAD

**Issue by Cadair** [opsdroid/opsdroid#719]
This is a proposal on how to resolve #536. It came from a discussion with @SolarDrew 

## Event Classes

Currently opsdroid provides a `Message` class which is designed to represent a simple textual message. This proposal is for a hierarchy of different types of events decended from an `Event` abstract class.

For example a simple inheritance chain might be `Event` > `Message` > `Image`.

Other types of events may include `RoomCreation` `RoomInvite` or `PullRequest`.

### Connector Specific Event Classes

To fully support the features of all the disperate platforms, I anticipate that connectors might want to implement derived event specific to their platform. Although I hope this might be rather rare. One example I could think of is `MatrixStateEvent` for arbitrary state keys in matrix rooms. (I.e. what the
matrix state database plugin uses.)

An issue I can not see a nice solution to is now to import these classes into skills if the connectors are loaded via the plugin system.

# Matchers and Events

If a connector emits an event of type `Image` a skill may want to handle it differently to an event of type `Message`. Therefore an event type matcher should be provided. Given that `Image` may be a subclass of `Message` the matcher which is the most specialised (longest walk down the MRU?) should be matched.

I am strongly in favor of having the skills import the event types from opsdroid and pass them to `@event_matcher` rather than using strings or other way of specifying type.

I also suggest that for the first version of this we only match on event type and not other properties.

### Connector Specific Event Matching

It is possible a skill will want to only match `PullRequest` or even `Message` events emitted by a specific connector. To facilitate this I propose that on creation of an event a `source` parameter is given, which would normally take the name of the emitting connector. This should allow both matching on connector while retaining a more generic event hierarchy in opsdroid core.

## Responding to Events

Currently messages are responded to by calling `Message.respond` which in turn calls `connector.respond`. We propose that a new method `opsdroid.send` be implemented for dispatching events. This method would have the following signature: (type hints for API design only)

```python
def send(event: Event, connector: str, target: str)
```

This method would be called by `Event.respond` which would pre-populate the `connector` and `target` keyword arguments based on the properties given when it was created. (Respond here meaning "respond in the same room (and therefore connector) as the original message).

Note: This API is specified as `connector: str` this is based on the idea that as part of this work the `opsdroid.connectors` property should be converted from a list to a dictionary. Therefore, making the `send` API take a string makes sense, because it would be effectively a shortcut for `connector=opsdroid.connectors["connector"]`. The `opsdroid.connectors` property being the only way I can see to get the currently configured connectors.

### Handling different events in the connector

The connectors should also implement a `send` method which would get an `Event` subclass and then they can process that event.

I haven't given this much thought, but something based around `functools.singledispatch` could work well, as we would be dispatching on type in the connector.

I.e. the base connector class could have a `send` method which is a generic function (with a modified version of singledispatch on it), then in the derived classes you could have a method like:

```python

    @send.register(Image)
    def send_image(event):
        pass
```

I am a little fuzzy on how singledispatch would work with classes and subclasses etc, so just take this as an idea!

**Comment by jacobtomlinson** [comment]
🎉 This all sounds excellent and right along the lines of what I was thinking of doing myself. 

I have a few questions that I would like your opinion on. There are not criticisms of this proposal, just questions which have come to mind while reading 😊. Some are relate to what you propose, some are related to opsdroid in general prompted by your proposal.

- Would `Event` and all subclasses have a `respond` method? Or perhaps just `Message` and all subclasses of that? Or something else?
- How would respond behave on non `Message` events like `Image`? Should it respond with text as it currently does? Should we modify it to take a new object like a `Message` or `Image` and respond with that? What happens if you try to respond with a `PullRequest` event object, that doesn't make sense.
- How should we handle natural language parsing? Currently `opsdroid.parse()` is called on `Message` objects by the connector in the `listen` loop. Do we want subclasses of `Message` to be parsed? Do we want other event types to be parsed (like `PullRequest` for example which has a textual element)? We could parse all events and check them for a `text` property and parse that.
  - To continue this thought I'm planning on adding _constraints_ soon which are a way of filtering matches. So we could have `@match_regex(r'some text')` and `@contrain_event(PullRequest)` to limit a skill to only be called on pull requests which match the regex. 
- Is the current way of sending messages to the connector an optimal user experience? Currently you either call the `respond` convenience method on a `Message` or construct a new `Message` and pass it to `connector.respond()`. Does `connector.respond()` even make sense as a method name any more, should it be `send` instead as you suggest?

**Comment by Cadair** [comment]
All these comments are very encouraging, I think we are all thinking along the same lines here.

> Would Event and all subclasses have a respond method? Or perhaps just Message and all subclasses of that? Or something else?

I would go for yes, all events should have a `respond` method. At the very least all events will have a connector property, and `target` (room) is currently (and presumably will continue to be) optional. In Drews example, of a room creation event, I would just expect there not to be a target property.


(Taking the next one out of order for clarity with my other responses)
> Does connector.respond() even make sense as a method name any more, should it be send instead as you suggest?

I think that we should remove the ability to send/respond on the connector at all (i.e. make it private API) in favour of implementing `opsdroid.send` (send being  a better method name like @SolarDrew says). I think that in a multi-connector universe having one send method on the main opsdroid instance makes more sense than having to first get the connector object.

> How would respond behave on non Message events like Image? Should it respond with text as it currently does? Should we modify it to take a new object like a Message or Image and respond with that? What happens if you try to respond with a PullRequest event object, that doesn't make sense.

I agree with @SolarDrew on this.

In my mind, `respond` is a shortcut for "send with connector and target from this event". I would expect it to be perfectly valid to respond with a `PullRequest` event, it is then up to the connector to decide if that is an event type it can handle, and if the target makes sense.

The example of `message.respond(PullRequest(...))` I think could actually work, I would think that in the case of a github connector, a comment on a PR or issue might result in a `Message` event being emitted, so I would expect that `Message` event to have a `target` equal to the issue it was a comment on i.e. `target="opsdroid/opsdroid#536"`. So `message.respond(PullRequest(...))` is equivalent to `opsdroid.send(PullRequest(...), connector="github", target="opsdroid/opsdroid#536")` the github connector could choose to interpret the `target` there not as an issue, but just as a repo (i.e. strip the issue number) and open a PR on that repo.

> How should we handle natural language parsing? [...] We could parse all events and check them for a text property and parse that. 

This makes sense to me. I think coming up with a set of properties for the events makes sense `text` being a good example of one.

> I was wondering if it might make more sense to pass additional conditions to the event matcher, so you can use just that rather than a constraint plus another matcher. 

I think I am slightly more in favour of the idea of stacking multiple criteria, each one in its own decorator, as @jacobtomlinson says. On the other hand, I think that what @SolarDrew says potentially allows matching on arbitrary properties of the event object, however, maybe that is not really what we want!!

**Comment by SolarDrew** [comment]
> I would just expect there not to be a target property.

I'm not sure I understand how this would play out in practice - where would the responding event go if there's no target? Or do you just mean no default target?

> (Taking the next one out of order for clarity with my other responses)

You're a damn renegade, @Cadair, they'll have your badge for this.

> allows matching on arbitrary properties of the event object, however, maybe that is not really what we want!!

I think we do though, at least in the long term. In the very general case where `Event`s can represent basically anything happening, I think we'll pretty quickly get past the point where there are a manageable number of things that could be matched against. The way I see it, working attribute matching into the event matcher makes the whole thing scale much more easily.

**Comment by jacobtomlinson** [comment]
The trouble with `@match_event(Message, text="some regex text")` is that you're making an assumption about the matcher you want to use (regex). You may be matching on a more complex intent model parser like Rasa NLU or something else.

**Comment by SolarDrew** [comment]
> you're making an assumption about the matcher you want to use (regex)

I'm not sure that necessarily needs to be the case. I think `text=` would match just text content and take the place of `@match_regex`, (and in fact, `regex=` would probably be clearer here) but you could also have an `intent=` keyword which would pass to Rasa NLU or whatever underneath, so you needn't lose that functionality.

**Comment by Cadair** [comment]
The issue I see is you have multiple properties of the event you might want to match (text, title, room, etc) and multiple ways you might want to match them (regex, equality, NLU etc).

**Comment by jacobtomlinson** [comment]
Referencing https://github.com/opsdroid/opsdroid/pull/721#discussion_r250925108

> Arguably, this whole [react] method should be deprecated in favour of some type of call to Message.send i.e.:
>
> `Message.send(Reaction("😼"))`

**Comment by Cadair** [comment]
I have been working on the next stage of the implementation of this, the `Connector.send` part. Take a look at this gist, where I have tested out my thoughts.

https://gist.github.com/Cadair/7f04936cd64112f9137e64bb380a26ab

**Comment by Cadair** [comment]
Bad stale bot!

**Comment by jacobtomlinson** [comment]
Can we get a status update on this? What needs to happen next for it to progress?

**Comment by Cadair** [comment]
I think #851 is the last major step, at least of the proof of concept. 

There is a lot of other stuff that could be done afterwards, i.e. constrains on matchers, as well as needing to make the other connectors emit events other than Message, but I would consider this closed by #851 I think.

**Comment by jacobtomlinson** [comment]
I had actually intended my comment to be on #851, not sure how it ended up here 😆
